<html><header><title>Notify on failed update</title></header><body><div style='float:right'><a href='http://cslanet.com'><img src='https://github.com/MarimerLLC/csla/raw/master/Support/Logos/csla%20win8_compact_s.png'/></a></div><p><h1>Notify on failed update</h1><span style='font-size:small'>Old forum URL: forums.lhotka.net/forums/t/9971.aspx</span></p><hr><div style='padding:0 15 3 15;background-color:powderblue'><h2>AaronH posted on Friday, January 21, 2011</h2><p>I have a businessListBase where I don&#39;t want the updates to succeed or fail as a transaction.&nbsp; If record 5 of 20 fails on an update, I want the updates (DP_Update)&nbsp;to continue.&nbsp; I then want to somehow invalidate record 5 so that when the list goes back to the UI, the user can be notified that the update for record 5 failed.</p>
<p>Does the CSLA intrinsically support this behavior, or will I have to create my own mechanism to support this (ie, setting a flag / error message propertie(s) on the object and having the UI display them based on those properties being set)?</p>
<p>Thanks!</p></div><div style='padding:0 15 3 15;background-color:lightgreen'><h2>ajj3085 replied on Friday, January 21, 2011</h2><p>A dyanmic list might be more of what you&#39;re looking for.&nbsp; It treats items in the collection as root objects.&nbsp; It&#39;s main purpose is to save the row as you leave it in the grid.</p></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>AaronH replied on Saturday, January 22, 2011</h2><p>Unfortunately I can&#39;t use that option as the user is allowed to select multiple records and make a change that applies to all of them.&nbsp; This would cause multiple server hits which I want to avoid.</p>
<p>In the interim, I&#39;m just checking a return condition on a stored proc and setting a generic &quot;record already updated by another user&quot; or &quot;record has been posted and cannot be changed&quot; on an object property&nbsp;itself (UpdateStatusMessage) that can be displayed in a grid cell.&nbsp; </p>
<p>I was really hoping to add a broken rule to the object and have the record highlighted in red in the bound grid when the objects came back from an update.&nbsp; Unfortunately that just caused more problems than I was willing to wrestle with.</p>
<p>I&#39;d be curious as to how others are handling a similar scenario.</p></div><div style='padding:0 15 3 15;background-color:lightgreen'><h2>RockfordLhotka replied on Saturday, January 22, 2011</h2><p>If you are using the Transactional attribute on your root method the update transaction will only rollback if an exception is allowed to flow up from the root method.</p>
<p>But really, it doesn&#39;t sound like you want a transaction to protect the set of updates, rather it sounds like you want to do one update at a time, each basically being its own transaction. Since SQL invocations are ACID, each one is protected individually automatically.</p>
<p>In other words, it sounds to me like you can (more or less) do what you want by</p>
<ol>
<li>Not using a transaction across the entire operation</li>
<li>Wrapping all child updates in a try..catch block to catch any exception</li>
<li>Breaking a rule in the child object to reflect the exception</li>
</ol>
<p>If I were doing this I&#39;d probably have each child do its own insert/update/deleteself, and that&#39;s where the try..catch would exist. That way you are in the child object when the failure occurs. I&#39;d then have a private property to store the exception or exception message. And I&#39;d have a business rule that is broken when this value is not null.</p>
<p>So when you catch an exception in the Child_Update (for example), you&#39;d set this private DataException property and then call CheckRules to run this rule, so the object is now InValid (or at least has a warning).</p>
<p>You&#39;ll probably need to override IsSavable and Save to allow saving of the object when it has this error. Or have some way to clear the DataException property after the user has change the object. In short - you need some way for the object to be &quot;valid&quot; once the user presumably fixes the problem.</p></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>AaronH replied on Saturday, January 22, 2011</h2><p>That&#39;s perfect.&nbsp; Overriding the IsSavable and Save() is just what i need in order to ignore the fact that the list is essentially invalid when it returns in this scenario (if the list is invalid upon return, no other changes to the list will be allowed to be saved, which is the issue i was facing earlier).</p>
<p>Thanks!</p></div><p style='font-size:small'>Copyright (c) Marimer LLC</body></html>
